v2/contribute/decisions/oauth/0014-not-expiring-tokens.mdx

```diff
@@ -0,0 +1,44 @@
+---
+id: "0014"
+title: Use long expiration times instead of non-expiring tokens
+hide_title: true
+---
+
+import DecisionHeader from "/src/components/decisionLogs/DecisionHeader"
+import ArgumentPro from "/src/components/decisionLogs/ArgumentPro"
+import ArgumentCon from "/src/components/decisionLogs/ArgumentCon"
+import ArgumentNeut from "/src/components/decisionLogs/ArgumentNeut"
+
+# Use long expiration times instead of non-expiring tokens
+
+<DecisionHeader status="proposed" lastUpdate="2023-05-11" created="2023-05-11" deciders={["rishabhpoddar", "porcellus"]} proposedBy={["porcellus"]} />
+
+## Context and Problem Statement
+
+In some M2M use-cases the user may require a token to never expire.
+
+## Considered Options
+
+* Recommend long expiration times instead of non-expiring tokens
+* Set the expiration time to null for non-expiring tokens
+
+## Decision Outcome
+
```

Comment:
This refers to access token expiry right? Does this also apply to refresh token expiry?

Reply:
If the access tokens are very login lived and the user wants to revoke them - how will they do it? Is there something in the access token which the user can check in the db to see if it's revoked or not - how can users check this on a code level (for example for our sessions, we have a `checkDatabase` boolean)?

Reply:
- This refers to the access token, but the refresh token expiration is also customizable.
- We expose a token verification endpoint that checks the database. In this case they have to basically handle these as opaque tokens.

Reply:
Sounds good. Please add these points to this ADR. Also mention that the expiry thing applies to refresh tokens as well - there will be a default config in the core for refresh token lifetime, but the sdk can specify a different value.

Also, once an access / refresh token pair is created, and when the refresh token is being used to generate new tokens - those tokens lifetime can't be changed for that session right? 

Also this has implications on the db structure - we will need to store the token's lifetime per token row in the db (since they can be different for different tokens) - @AdithyaSSup please see

Reply:
I think allowing users to change the lifetime of the refreshed token is OK. Not sure about use-cases, but it's possible that someone would require a session "upgrade" where after some extra validation (e.g.: MFA). Not sure if anybody requires this, but I don't see an issue of making this configurable when calling refresh either.
Keep in mind, that this will not be exposed to the frontend and in most cases will not even be used by the backend (which will fall back to the default set in the core config.)

Reply:
Oh. I'm ok with the lifetimes being fixed after creation (i.e, during refresh they can't be set to a new lifetime). What i was saying is that since users can set custom lifetimes during creation of the tokens, how will we know what those values are during the next refresh if we don't store this value in the db?

Reply:
What I'm saying is that we don't really need to know - the user can just pass it the next time around.
Manual refreshing the token by directly calling the core is not really an intended use-case.
The refresh endpoint/functionality exposed by the backend SDK will:
1. Build an access and id token payloads based on the user/session id and the scopes (and context)
2. Pass the above payloads to the user provided callbacks if present
3. Get the token lifetimes from a user provided callback if present (otherwise falling back to the value set in the core config)
4. call the core, which will receive both the custom payloads and lifetimes (so there is no need to store them)

Keep in mind, that this is not really intended to be used the same way as our own access tokens: if you are changing the state of the user in the auth server we have no way to update these tokens for multiple reasons, so I think it's better to regenerate the tokens during refresh calls and (matching this pattern) re-generating the lifetimes as well.

Let's discuss this on a call tomorrow, then I'll add it to an ADR.

Reply:
- Add a new core endpoint to fetch the token info based on the refresh (this will be used as an input to the payload/lifetime generator callbacks)
- Add a new column into the token info if we require a client secret when refreshing (set to true if the auth code/token was created with a client secret)
- The client secret should be required to consume an authcode if not using PKCE (thus setting the above field to true in this case)
- We are OK with the user not  being able to disallow offline access (== issuing refresh tokens) on a client level in the core. They can do this by overriding functions in the backend SDK.